### PR TITLE
fluids: Fix Riemann solvers *_fwd mode functions

### PR DESCRIPTION
**Pull request by jrwrigh**
This verifies and fixes the `*_fwd` functions associated with the Riemann solver.

This also adds several tests for verifying those `*_fwd` functions as well. They can be run via `-riemann_solver_unit_tests`.

Note, however this does **not** get quadratic convergence from the non-linear solver, while using `-snes_fd` _does_ get quadratic convergence. Example with the finite difference Jacobian:
```
make build/fluids-navierstokes -j && build/fluids-navierstokes -ceed /cpu/self -options_file examples/fluids/gaussianwave.yaml -state_var primitive -ts_max_steps 2 -ts_monitor -snes_monitor -stab none -snes_fd true -snes_rtol 1e-6 -snes_lag_jacobian 1 -dm_plex_box_faces 20,20,1

0 TS dt 0.002 time 0.
    0 SNES Function norm 1.169470031986e+00
    1 SNES Function norm 1.059833797585e-01
    2 SNES Function norm 2.812942713581e-04
    3 SNES Function norm 1.138812349517e-06
    0 SNES Function norm 1.405334647329e+00
    1 SNES Function norm 8.875162078750e-02
    2 SNES Function norm 2.600321862598e-04
    3 SNES Function norm 1.145273173210e-06
    0 SNES Function norm 2.338940063972e+00
    1 SNES Function norm 2.119338056945e-01
    2 SNES Function norm 5.627468083189e-04
    3 SNES Function norm 2.278664021262e-06
1 TS dt 0.002 time 0.002
    0 SNES Function norm 4.190564303475e+00
    1 SNES Function norm 1.587727618966e-01
    2 SNES Function norm 6.243533944119e-04
    3 SNES Function norm 2.780230682175e-06
```

and without:
```
make build/fluids-navierstokes -j && build/fluids-navierstokes -ceed /cpu/self -options_file examples/fluids/gaussianwave.yaml -state_var primitive -ts_max_steps 2 -ts_monitor -snes_monitor -stab none -snes_fd false -snes_rtol 1e-6 -snes_lag_jacobian 1 -dm_plex_box_faces 20,20,1

0 TS dt 0.002 time 0.
    0 SNES Function norm 1.169470031986e+00
    1 SNES Function norm 1.059669236937e-01
    2 SNES Function norm 2.967002852497e-03
    3 SNES Function norm 5.026261931934e-05
    4 SNES Function norm 3.367308181466e-06
    5 SNES Function norm 2.819682770245e-07
    0 SNES Function norm 1.405334744560e+00
    1 SNES Function norm 8.875161734422e-02
    2 SNES Function norm 2.570679036615e-03
    3 SNES Function norm 4.640260451592e-05
    4 SNES Function norm 2.242670488445e-06
    5 SNES Function norm 2.199217165059e-07
    0 SNES Function norm 2.338940063972e+00
    1 SNES Function norm 2.119338473873e-01
    2 SNES Function norm 5.934005704992e-03
    3 SNES Function norm 1.005252386391e-04
    4 SNES Function norm 6.734616365045e-06
    5 SNES Function norm 5.639365544084e-07
1 TS dt 0.002 time 0.002
    0 SNES Function norm 4.190564676279e+00
    1 SNES Function norm 1.587727962362e-01
    2 SNES Function norm 5.111016681575e-03
    3 SNES Function norm 1.112404445653e-04
    4 SNES Function norm 3.654018550379e-06
```

### CODE REVIEW

**Comment by knepley** [comment]
@jrwrigh I am not trying to make work for you. However, if you want to fix the Riemann Jacobian routines, PETSc has nice support. If you use `-snes_test_jacobian -snes_test_jacobian_view` it will print out the analytic Jacobian, the FD Jacobian, and the difference. The difference makes it easy to see what terms are wrong.

**Comment by jrwrigh** [comment]
Interestingly, using:
```
build/fluids-navierstokes -ceed /cpu/self -options_file examples/fluids/gaussianwave.yaml -state_var primitive -ts_max_steps 1 -ts_monitor -snes_monitor -stab none -snes_rtol 1e-6 -dm_plex_box_faces 2,2,1 -snes_lag_jacobian 1 -dm_plex_box_bd 'none,none,periodic' -bc_symmetry_z -dm_plex_shape zbox -snes_test_jacobian 1e-8 -snes_test_jacobian_view
```

I'm seeing no difference in the reported Jacobian difference between `main` and this branch. I suppose that shouldn't be too surprising since in the first time step of the problem, the Riemann boundary condition should be solving the trivial problem (interior state identical to freestream state). So it appears the issue lies elsewhere.

That said, these fixes are still fixes as verified by the added unit tests.

File: examples/fluids/qfunctions/riemann_solver.h [hunk 1]
```diff
@@ -84,13 +84,11 @@ CEED_QFUNCTION_HELPER StateConservative Flux_HLL_fwd(State left, State right, St
...
-    const CeedScalar S_diff      = S_r - S_l;
-    const CeedScalar F_hll_denom = S_r * F_l[i] - S_l * F_r[i] + S_l * S_r * U_diff;
+    const CeedScalar S_diff = S_r - S_l;
 
-    dF_hll[i] += ((F_l[i] + S_r * U_diff) * S_diff - F_hll_denom) / Square(S_diff) * dS_r;
-    dF_hll[i] += ((-F_r[i] + S_r * U_diff) * S_diff + F_hll_denom) / Square(S_diff) * dS_l;
-    dF_hll[i] += (S_r * dF_l[i] - S_l * dF_r[i] + S_r * S_l * dU_r[i] - S_r * S_l * dU_l[i]) / S_diff;
+    dF_hll[i] += (S_l * (-F_l[i] + F_r[i] + S_l * U_l[i] - S_l * U_r[i]) / Square(S_diff)) * dS_r;
+    dF_hll[i] += (S_r * (F_l[i] - F_r[i] - S_r * U_l[i] + S_r * U_r[i]) / Square(S_diff)) * dS_l;
+    dF_hll[i] += (S_r * dF_l[i] - S_l * dF_r[i] + S_r * S_l * (dU_r[i] - dU_l[i])) / S_diff;
```
Comment (jeremylt) [inline examples/fluids/qfunctions/riemann_solver.h:91]: is this the part that had an error?

Comment (jrwrigh) [reply examples/fluids/qfunctions/riemann_solver.h:91]: There were two parts, both in the same file. So this is one of them.

Comment (jrwrigh) [reply examples/fluids/qfunctions/riemann_solver.h:91]: The other error (for `a_roe` in `ComputeHLLSpeeds_Roe_fwd`) is easy to spot the difference. TBH I'm not sure about the difference here. But the fix comes from `sympy` and allowed the test to pass (it wasn't before this fix).

Comment (jeremylt) [reply examples/fluids/qfunctions/riemann_solver.h:91]: Cool, just wanted to make sure I was understanding which parts were test changes, convention changes, and bug fixes

**Review by jeremylt — APPROVED** [review]
